File: Java/2351-2400/2385. Amount of Time for Binary Tree to Be Infected.java
// https://leetcode.com/problems/amount-of-time-for-binary-tree-to-be-infected/

// https://leetcode.com/problems/amount-of-time-for-binary-tree-to-be-infected/solutions/4538708/o-n-java-python-javascript-typescript-c-c-binary-tree/

class Solution {
    private int result;

    public int amountOfTime(TreeNode root, int start) {
        DFS(root, start);
        return result;
    }

    private int DFS(TreeNode node, int start){
        if(node == null) return 0;

        int leftDepth = DFS(node.left, start);
        int rightDepth = DFS(node.right, start);

        if(node.val == start){
            result = Math.max(leftDepth, rightDepth);
            return -1;
        }
        else if(leftDepth >= 0 && rightDepth >= 0)
            return Math.max(leftDepth, rightDepth)+1;
        
        result = Math.max(result, Math.abs(leftDepth - rightDepth));
        return Math.min(leftDepth, rightDepth) - 1;
    }
}

File: Python/2351-2400/2385. Amount of Time for Binary Tree to Be Infected.py
# https://leetcode.com/problems/amount-of-time-for-binary-tree-to-be-infected/

# https://leetcode.com/problems/amount-of-time-for-binary-tree-to-be-infected/solutions/4538708/o-n-java-python-javascript-typescript-c-c-binary-tree/

class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        result = 0

        def DFS(node, start):
            if node == None:
                return 0

            leftDepth = DFS(node.left, start)
            rightDepth = DFS(node.right, start)

            if node.val == start:
                Solution.result = max(leftDepth, rightDepth)
                return -1
            
            elif leftDepth >= 0 and rightDepth >= 0:
                return max(leftDepth, rightDepth)+1
            
            Solution.result = max(Solution.result, abs(leftDepth - rightDepth))
            return min(leftDepth, rightDepth) - 1

        DFS(root, start)
        return Solution.result    
